### DISCUSSION TITLE
How to unit test a transaction

### DISCUSSION THREAD

**Discussion by bensullivan-ag** [prisma/prisma#14066] [Q&A] [answered]
Hi

I wanted to write a unit test for a transaction.  Do you have an example of this anywhere?

Otherwise, how should a prisma client operation be mocked to simulate a transaction failure and rollback?

Thanks

Ben

**Comment by ruheni** [comment, marked as answer]
Hey 👋🏽

We don't have documentation on how to unit test for a transaction in our docs, but we have [example unit tests](https://www.prisma.io/docs/guides/testing/unit-testing#example-unit-tests), which are extensive. Have you tried it out with transactions, and what challenges are you experiencing?  🙂

**Reply by MichaelMcKavanagh** [reply]
Unfortunately the original question has not been answered, and I arrived here with the same problem.  The prisma.$transaction is completely skipped by the unit test (including everything inside it).  I believe it's being 'automocked'.  Basically need a way to not have it automocked, so that the code within the prisma.$transaction can be tested.  Any help as to how to have this happen or any comprehensive examples would be much appreciated.  Thanks.

**Reply by pantharshit00** [reply]
If it is a single test, following might work depending on a usecase:
```ts
    jest.spyOn(prisma, '$transaction').mockImplementation((callback) => {
      return callback(prisma)
    })
```
and then mock methods on regular prisma objects. Caveats:
1. It won’t be running in actual transaction, so if anything goes wrong no changes will be rolled back. But I assume the goal is to avoid actual DB anyway, otherwise, mocking would’n have came up
2. Since we are passing actual prisma instance to the tx callback, it can do more than an actual transaction object will be able to: bunch of methods are not available on real tx proxies. TS will be able to prevent it, but there will be no corresponding runtime check

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
